cmd/src/debug_common.go

```diff
@@ -0,0 +1,470 @@
+// and context needed for later archive functions. Don't forget to defer close on these
+// after calling setupDebug!
+func setupDebug(base string) (*os.File, *zip.Writer, context.Context, error) {
+	// open pipe to output file
+	out, err := os.OpenFile(base, os.O_CREATE|os.O_RDWR|os.O_EXCL, 0666)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to open file: %w", err)
+	}
+	// increase limit of open files
+	err = setOpenFileLimits(64000)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to set open file limits: %w", err)
+	}
+	// init zip writer
+	zw := zip.NewWriter(out)
+	// init context
+	ctx := context.Background()
+
+	return out, zw, ctx, err
+}
+
+// TODO: Currently external services and site configs are pulled using the src endpoints
+
+// getExternalServicesConfig calls src extsvc list with the format flag -f,
+//and then returns an archiveFile to be consumed
+func getExternalServicesConfig(ctx context.Context, baseDir string) *archiveFile {
+	const fmtStr = `{{range .Nodes}}{{.id}} | {{.kind}} | {{.displayName}}{{"\n"}}{{.config}}{{"\n---\n"}}{{end}}`
+
+	f := &archiveFile{name: baseDir + "/config/external_services.txt"}
+	f.data, f.err = exec.CommandContext(ctx, os.Args[0], "extsvc", "list", "-f", fmtStr).CombinedOutput()
+
+	return f
+}
+
+// getSiteConfig calls src api -query=... to query the api for site config json
+// TODO: correctly format json output before writing to zip
+func getSiteConfig(ctx context.Context, baseDir string) *archiveFile {
+	const siteConfigStr = `query { site { configuration { effectiveContents } } }`
+	f := &archiveFile{name: baseDir + "/config/siteConfig.json"}
+	f.data, f.err = exec.CommandContext(ctx, os.Args[0], "api", "-query", siteConfigStr).CombinedOutput()
+
+	return f
+}
+
+/*
+Kubernetes stuff
+TODO: handle namespaces, remove --all-namespaces from get events
+TODO: have a confirmation step that warns user about which namespace and context they have kubectl pointed at
+*/
+
+type podList struct {
+	Items []struct {
+		Metadata struct {
+			Name string
+		}
+		Spec struct {
+			Containers []struct {
+				Name string
+			}
+		}
+	}
+}
+
+// Run kubectl functions concurrently and archive results to zip file
+func archiveKube(ctx context.Context, zw *zip.Writer, verbose, configs bool, namespace, baseDir string, pods podList) error {
+	// Create a context with a cancel function that we call when returning
+	// from archiveKube. This ensures we close all pending go-routines when returning
+	// early because of an error.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// setup channel for slice of archive function outputs, as well as throttling semaphore
+	ch := make(chan *archiveFile)
+	wg := sync.WaitGroup{}
+	semaphore := semaphore.NewWeighted(8)
+
+	// create goroutine to get kubectl events
+	wg.Add(1)
+	go func() {
+		if err := semaphore.Acquire(ctx, 1); err != nil {
+			// return err
```

Comment:
I think you can just return early in these conditions; an error is only returned when the context is canceled so there's no one to read the error.

Reply:
returning early 👍